pkg/dynamicaccess/history.go

```diff
@@ -1,36 +1,158 @@
 package dynamicaccess
 
 import (
-	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethersphere/bee/v2/pkg/kvs"
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+	"time"
+
+	"github.com/ethersphere/bee/v2/pkg/file"
+	"github.com/ethersphere/bee/v2/pkg/manifest"
+	"github.com/ethersphere/bee/v2/pkg/manifest/mantaray"
+	"github.com/ethersphere/bee/v2/pkg/swarm"
 )
 
 type History interface {
-	Add(timestamp int64, kvs kvs.KeyValueStore) error
-	Get(timestamp int64) (kvs.KeyValueStore, error)
-	Lookup(at int64) (kvs.KeyValueStore, error)
+	Add(ctx context.Context, actRef swarm.Address, timestamp *int64) error
+	Lookup(ctx context.Context, timestamp int64, ls file.LoadSaver) (swarm.Address, error)
+	Store(ctx context.Context) (swarm.Address, error)
 }
 
 var _ History = (*history)(nil)
 
+var ErrEndIteration = errors.New("end iteration")
+
 type history struct {
-	history map[int64]*kvs.KeyValueStore
+	manifest *manifest.MantarayManifest
+}
+
+func NewHistory(ls file.LoadSaver, ref *swarm.Address) (*history, error) {
+	var err error
+	var m manifest.Interface
+
+	if ref != nil {
+		m, err = manifest.NewDefaultManifestReference(*ref, ls)
+	} else {
+		m, err = manifest.NewDefaultManifest(ls, false)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	mm, ok := m.(*manifest.MantarayManifest)
+	if !ok {
+		return nil, fmt.Errorf("expected MantarayManifest, got %T", m)
+	}
+
+	return &history{manifest: mm}, nil
 }
 
-func NewHistory(topic []byte, owner common.Address) *history {
-	return &history{history: make(map[int64]*kvs.KeyValueStore)}
+func (h *history) Manifest() *manifest.MantarayManifest {
+	return h.manifest
 }
 
-func (h *history) Add(timestamp int64, kvs kvs.KeyValueStore) error {
+func (h *history) Add(ctx context.Context, actRef swarm.Address, timestamp *int64) error {
```

Comment:
maybe use the same timestamp <= 0 check as in LookpUp

Reply:
This also makes me think. Is a validation like this correct in a fn? Shouldn't we move all of this validation one layer above?